### PR TITLE
ENH: Read filelike objects

### PR DESCRIPTION
**Pull request by Sangarshanan**
Tried to support reading of file-like objects using geopandas, looking for feedback and suggestions 
Ref #388 and #1323

### CODE REVIEW

File: geopandas/io/file.py [hunk 1]
```diff
@@ -34,15 +35,15 @@ def _is_url(url):
         return False
 
 
-def read_file(filename, bbox=None, mask=None, rows=None, **kwargs):
+def read_file(filepath_or_buffer, bbox=None, mask=None, rows=None, **kwargs):
```
Comment (martinfleis) [inline geopandas/io/file.py]: We certainly can't change the name of argument as it would break backward compatibility.

**Comment by martinfleis** [comment]
Thanks for the PR! This looks good. What are all the possible options we might want to support (and test)? Drawing from linked issues:

- [ ] tempfile.TemporaryFile (is it supported now? can we support it and test it?)
- [x] pathlib.Path
- [x] String IO
- [x] binary IO
- [x] raw IO
- [x] Bytes IO

Anything else to support?

File: geopandas/io/file.py (hunk 1 above)
Comment (Sangarshanan) [reply geopandas/io/file.py]: Got it... I actually changed it to emulate pandas

**Comment by Sangarshanan** [comment]
Yup it is possible to read `tempfile.TemporaryFile `added a test for it, also it is actually possible to read any object with a read() method now. Things I see missing in the test are py.path and s3 objects, url's. But py.path doc says that it's in maintenance mode and has asked to refer pathlib for path operations. I am also checking out reading s3 urls (possible with fiona) and even streaming s3 objects ref: https://github.com/pandas-dev/pandas/pull/22520 but maybe that is a separate issue 

Should I address the user guide changes here?

File: geopandas/io/file.py [hunk 2]
```diff
@@ -35,17 +35,18 @@ def _is_url(url):
...
 
     .. versionadded:: 0.7.0 mask, rows
 
     Parameters
     ----------
-    filepath_or_buffer: str
+    filename: str, path object or file-like object
         Either the absolute or relative path to the file or URL to
-        be opened.
+        be opened, or any object with a read() method(such as an open file
```
Comment (martinfleis) [inline geopandas/io/file.py]:
```suggestion
        be opened, or any object with a read() method (such as an open file
```

**Comment by Sangarshanan** [comment]
I have added some basic content to the user guide, not sure if should I add more detailed examples explaining every file like object supported?

**Comment by martinfleis** [comment]
I don't think we need to add more details. Docstring clearly indicates that there are other options apart from file path.

I have pushed some minor style cleaning. I recommend using pre-commit hook to make sure that the code follows `black` and `flake8` - https://geopandas.readthedocs.io/en/latest/contributing.html#style-guide-linting

Thanks!

**Comment by martinfleis** [comment]
Thank you @Sangarshanan!

**Pull request merged**